libnmstate/nm/connection.py

```diff
@@ -21,26 +21,67 @@
 from . import nmclient
 
 
-def create_profile(settings):
-    con_profile = nmclient.NM.SimpleConnection.new()
-    for setting in settings:
-        con_profile.add_setting(setting)
+class ConnectionProfile(object):
+
+    def __init__(self, profile=None):
```

Comment:
IMHO the options to create a ConnectionProfile should be expressed in the constructor and not in extra create or import methods. When creating a ConnectionProfile with profile=None (the default), then commit() and update() and add() will most likely not work or in other words, the object is not properly initialized.

Reply:
The object can be in a state that some of its actions are not valid.
I could convert this constructor argument to an import_by_profile if you prefer.

Expressing how the object should be created through the arguments is ugly, especially when the combination of the input may or may not be valid due to collisions. I think that explicitly describing the steps through methods is much better and clearer. it also allows the object to expand in a nicer manner (through additional actions on it).

Reply:
The current way is uglier than having multiple arguments in the constructor and it is also misleading that a constructor does not create the actual object but that there is a separate create method. @cathay4t, what do you think?

IMHO the other constructors should be at least class methods that return a proper object. But actually create() is IMHO just the constructor and the import_by methods actually belong in other classes so that a Device object returns a ConnectionProfile or the client can be used to look up a ConnectionProfile based on the ID.

Reply:
It is a common practice to create "unloaded" objects and it allows better maintenance and extendibility options.
At the minimum, it is certainly not wrong.

A constructor with unused arguments is wrong.
Functions with unused arguments (in an OR relationship) are also wrong, unfortunately we do have them as well (although I consider constructor more important in this regard).

This PR is proposing one way of abstracting things. I am sure there are other ways to do it.
Up to now I was not convinced that this approach is not good or that it cannot be a base for future refactoring, but if you have something better, please suggest it through a PR so we can discuss it (can be a new PR or over this one).
(except of not agreeing on the approach, I also do not see how this current work can be adjusted to fit with your proposals)

Reply:
Having default values for arguments including constructors is best practice in Python and commonly used, see for example: https://docs.python.org/3.4/library/subprocess.html#subprocess.Popen

Reply:
These are not defaults, we are talking about multiple arguments which sometime collide between themself.
I do not think that having a constructor with tens of arguments and with potentially new introduced ones in the future is a good thing.